### ISSUE TITLE
Rule proposal: A Type-aware version of `no-restricted-syntax`

### ISSUE THREAD

**Issue by Woodz** [typescript-eslint/typescript-eslint#8650]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

eslint base `no-restricted-syntax` rule is very useful for restricting specific JS language features or code, but is not aware of TS types. I propose adding a similar rule that augments the ESTree AST with TS typing so that developers can restrict specific syntax that requires type awareness

### Fail Cases

```TypeScript
class MyClass {
  foo(): void {
   // This is dangerous and should not be used
   throw new Error();
  }
}
class MyOtherClass {
  foo(): void {
  }
}

const myClass = new MyClass();
// I want to use `no-restricted-syntax` to catch this call 
myClass.foo();

const myOtherClass = new MyOtherClass();
// But this call should be fine
myOtherClass.foo();
```

### Pass Cases

```TypeScript
N/A
```

**Comment by bradzacher** [comment]
On the surface this seems like an interesting idea to explore.
However when you delve more deeply it falls apart as a concept pretty quickly.

The point of `no-restricted-syntax` is allowing a codebase to quickly configure certain patterns that are to be disallowed. Syntax is easy to do here because it's trivially covered by ESQuery selectors. They're deterministic, easy to write, and easy to understand.
Put another way - you can easily express the constraints imposed by `no-restricted-syntax` with a single line - usually only a few dozen characters in a string - and it'll handle _all_ required cases.

OTOH types are not really any of those things. They get very messy very quickly due to TS's structural typing model, as well as union and intersection types.
For example you say `MyClass.foo` should be disallowed - but what about if you have `MySubClass extends MyClass` but it doesn't override `.foo`? Well logically that should also be blocked as well, right?
What about if the type is `MyClass & { prop: string }`? Well the `.foo` should be banned on that as well, right?
Or what about if the type is `MyClass | MyOtherClass` - should `.foo` be banned on that?

There are a lot of edge cases to consider and a lot of weirdness to deal with.

In other words - in contrast to a syntactical restriction you cannot represent a type restriction with a simple query.

---

Instead we recommend delving into building a custom rule here to encapsulate all of the required logic and also so that you can write tests to validate the complexities and edge cases.
https://typescript-eslint.io/developers/custom-rules

You can freely create a plugin locally for your project without publishing it - with flat configs it's now easier than ever!

**Comment by JoshuaKGoldberg** [comment]
Agreed. I personally would _love_ to have a version of this rule, but it's a very unexplored space that would need a lot of exploration before it makes its way into our stable plugin.

The main blocker that comes to mind for me is that there's no standard [ESQuery](https://github.com/estools/esquery)/[tsquery](https://github.com/phenomnomnominal/tsquery) equivalent for types in the TypeScript dev community. If there was, then we'd be able to use it in a rule like this. But there isn't and we're not well-positioned to make one.

Closing for now - but if the world changes and some kind of "types query" standard does exist, we'd be happy to revisit. Cheers!
